### DISCUSSION TITLE
Flat bands in band diagram calculation

### DISCUSSION THREAD

**Discussion by YueJiang2048** [NanoComp/meep#2003] [Q&A]
Hello,
I implement a GyrotropicDrude material for band diagram calculation. The resulted band diagram involves flat bands. 
Below is the result for a simple test calculation for a 2D square lattice with Drude material (non-gyrotropic) in a circle, using the mp.DrudeSusceptibility function, 
1) I found the number of flat bands is related to the resolution. Higher-resolution leads to more flat bands. 
2) In some of the regions, e.g., below 0.4 or between 0.7-0.9, the band diagram is clear, but other regions have those flat bands.

My question is, 
1) How to understand the flat bands 
2) How to reduce the flat band to get a clearer band diagram?

Thank you for any thoughts and suggestions!

The python codes are in the bottom. I attempt resolution = 30 and 16, and got the following band diagrams:

## Resolution = 30
![Screen Shot 2022-03-16 at 8 05 08 PM](https://user-images.githubusercontent.com/101615840/158711538-0fecd9f7-0d6d-4558-bb5b-6b0639d1fcf1.png)

## Resolution = 16
![Screen Shot 2022-03-16 at 8 06 44 PM](https://user-images.githubusercontent.com/101615840/158711631-47dd1e29-4550-42a5-be9c-877c59037a19.png)

## Overlapping of Res = 16 & Res = 30 plot
![drudeplt](https://user-images.githubusercontent.com/101615840/158713481-e87d519b-a89b-4363-bd38-a99a6233bde3.png)

```python
import meep as mp
import numpy as np
from matplotlib import pyplot as plt

# Some parameters to describe the geometry:
a = 1    # width of waveguide
r = 0.3   # radius of holes

# The cell dimensions
sx = 1;
sy = 1;
cell = mp.Vector3(sx,sy);

# Matierial
##Drude medium
susc = [mp.DrudeSusceptibility(frequency=1, gamma=0, sigma=1)]
mat = mp.Medium(epsilon=1,mu=1, E_susceptibilities=susc)

geometry = [mp.Cylinder(radius=r,
                        center=mp.Vector3(0,0,0),
                        material=mat)
           ]

resolution=16
fcen = 0.5  # pulse center frequency
df = 1  # pulse freq. width: large df = short impulse

src = [mp.Source(src=mp.GaussianSource(fcen, fwidth=df), component=mp.Hz,
              center=mp.Vector3(0.1034,-0.05342))]
sim = mp.Simulation(cell_size=cell,
                    geometry=geometry,
                    sources=src,
                    resolution=resolution,
                   ensure_periodicity=True)
kx = 0.8
sim.k_point = mp.Vector3(kx)

# k-path
Gamma = mp.Vector3()
X = mp.Vector3(0.5, 0.0, 0.0)
M = mp.Vector3(0.5, 0.5, 0.0)
 
Nk = 10 # number of points to interpolate between each k-point
kvs = mp.interpolate(Nk, [Gamma, X, M, Gamma])
all_freqs = sim.run_k_points(300,kvs);

fig = plt.figure(dpi=100,figsize=(5,5));
for i in range(len(all_freqs)):
    for ii in range(len(all_freqs[i])):
        plt.scatter(i,np.real(all_freqs[i][ii]),color='b',marker='.')




```

**Comment by nolanloo** [comment]
Did you figure it out? I have almost the same problem.

**Comment by YueJiang2048** [comment]
No..I didn’t. These flat bands seems to be intrinsic…however, it does not converge with resolution, which might mean that MEEP is not a good choice for the problem.
Best
Yue

From: Yang Lu ***@***.***>
Date: Monday, October 3, 2022 at 11:48 PM
To: NanoComp/meep ***@***.***>
Cc: YueJiang2048 ***@***.***>, Author ***@***.***>
Subject: Re: [NanoComp/meep] Flat bands in band diagram calculation (Discussion #2003)

Did you figure it out? I have almost the same problem.
